develop/blockchains/custom-blockchains/benchmarking.md

```diff
@@ -1,4 +1,223 @@
+
+## Running Benchmarks
+
+You can now compile your runtime with the following command with the `runtime-benchmarks` feature flag. It is crucial that you have a runtime with this flag enabled, as this is used for benchmarking later:
+
+```bash
+cargo build --features runtime-benchmarks --release
+```
+
+Once it is compiled with the correct feature set, you can run the benchmarking tool:
+
+```sh
+./target/release/<node-binary-name> benchmark pallet \
+--runtime <path-to-wasm-runtime> \
+--pallet <name-of-the-pallet> \
+--extrinsic '*' \
+--steps 20 \
+--repeat 10 \
+--output weights.rs
+```
+
+- `--runtime` - the path to your runtime's Wasm
+- `--pallet` - the name of the pallet you wish to benchmark. This pallet must be configured in your runtime and defined in `define_benchmarks`
+- `--extrinsic` - which extrinsic to test. Using `*` implies all extrinsics will be benchmarked
+- `--output` - where the output of the auto-generated weights will reside
+
+The result should be a `weights.rs` file containing the types you can use to annotate your extrinsic with the correctly balanced weights in your runtime. The output should be similar to the following. Some output has been omitted for brevity: 
+
+```sh
+2024-10-28 11:07:25 Loading WASM from ./target/release/wbuild/educhain-runtime/educhain_runtime.wasm
+2024-10-28 11:07:26 Could not find genesis preset 'development'. Falling back to default.
+2024-10-28 11:07:26 assembling new collators for new session 0 at #0
+2024-10-28 11:07:26 assembling new collators for new session 1 at #0
+2024-10-28 11:07:26 Loading WASM from ./target/release/wbuild/educhain-runtime/educhain_runtime.wasm
+Pallet: "pallet_parachain_template", Extrinsic: "do_something", Lowest values: [], Highest values: [], Steps: 20, Repeat: 10
+
+# ... Storage Info, Median Slopes, and Min Squares Analysis ...
+
+Created file: "weights.rs"
+2024-10-28 11:07:27 [  0 % ] Starting benchmark: pallet_parachain_template::do_something
+2024-10-28 11:07:27 [ 50 % ] Starting benchmark: pallet_parachain_template::cause_error
+```
+
+### Adding Benchmarked Weights to Pallet
+
+Once the `weights.rs` is generated, you may add the generated weights to your pallet. It is common that `weights.rs` become part of your pallet's root in `src/`:
+```rust
+use crate::weights::WeightInfo;
+
+/// Configure the pallet by specifying the parameters and types on which it depends.
+#[pallet::config]
+pub trait Config: frame_system::Config {
+    /// A type representing the weights required by the dispatchables of this pallet.
+    type WeightInfo: WeightInfo;
+}
+```
+
+After which, you may add this to the `#[pallet::weight]` annotation in the extrinsic via the `Config`: 
+
+```rust hl_lines="2"
+#[pallet::call_index(0)]
+#[pallet::weight(T::WeightInfo::do_something())]
+pub fn do_something(origin: OriginFor<T>, bn: u32) -> DispatchResultWithPostInfo { ... }
```

Comment:
Is the second parameter needed here? maybe we can get rid of it to make the code snippet even clearer